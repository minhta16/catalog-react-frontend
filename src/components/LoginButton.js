import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { Button, Avatar, Menu, MenuItem, Grid, Typography } from '@material-ui/core';

export class LoginButton extends Component {
  state = {
    anchorEl: undefined,
  };

  handleClickAvatar = (e) => {
    this.setState({
      anchorEl: e.target,
    });
  };

  handleCloseMenu = () => {
    this.setState({
      anchorEl: undefined,
    });
  };

  render() {
    const { onClick, variant, color, currentUser } = this.props;
    const { anchorEl } = this.state;
    const menu = (
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={this.handleCloseMenu}
        getContentAnchorEl={null}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        transformOrigin={{ vertical: 'top', horizontal: 'center' }}
      >
        <MenuItem onClick={this.handleCloseMenu}>Profile</MenuItem>
        <MenuItem onClick={this.handleCloseMenu}>Logout</MenuItem>
      </Menu>
    );
    return (
      <div>
        {Object.keys(currentUser).length ? (
          <Grid container alignItems="center">
            <Typography variant="body1">{`Welcome, ${currentUser.username}`}</Typography>
            <Avatar
              style={{ backgroundColor: '#fdb600', color: '#fff', margin: '0.5rem' }}
              onClick={this.handleClickAvatar}
            >
              {currentUser.username.charAt(0)}
            </Avatar>
            {menu}
          </Grid>
        ) : (
          <Button id="login-button" onClick={onClick} variant={variant} color={color}>
            Login
          </Button>
        )}
      </div>
    );
  }
}

LoginButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  variant: PropTypes.string,
  color: PropTypes.string,
  currentUser: PropTypes.object.isRequired,
};

LoginButton.defaultProps = {
  variant: 'contained',
  color: 'primary',
};

export default LoginButton;
