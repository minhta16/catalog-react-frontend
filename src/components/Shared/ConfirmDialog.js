import React from 'react';
import PropTypes from 'prop-types';
import {
  Dialog,
  DialogTitle,
  DialogContent,
  Typography,
  Button,
  DialogActions,
} from '@material-ui/core';

/**
 * This component creates a confirm dialog which calls onConfirm when the user clicks confirm and onCancel when the user clicks cancel
 */
export function ConfirmDialog(props) {
  const { title, contentText, open, onCancel, onClose, onConfirm } = props;
  return (
    <Dialog open={open} onClose={onClose}>
      <DialogTitle>{title}</DialogTitle>
      <DialogContent>
        <Typography variant="body1">{contentText}</Typography>
      </DialogContent>
      <DialogActions>
        <Button id="confirm-dialog-confirm" onClick={onConfirm} color="primary">
          Confirm
        </Button>
        <Button id="confirm-dialog-cancel" onClick={onCancel}>
          Cancel
        </Button>
      </DialogActions>
    </Dialog>
  );
}

ConfirmDialog.propTypes = {
  title: PropTypes.string,
  contentText: PropTypes.string,
  open: PropTypes.bool,
  onCancel: PropTypes.func,
  onConfirm: PropTypes.func,
  onClose: PropTypes.func,
};

ConfirmDialog.defaultProps = {
  title: '',
  contentText: '',
  open: false,
  onCancel: () => {},
  onConfirm: () => {},
  onClose: () => {},
};

export default ConfirmDialog;
