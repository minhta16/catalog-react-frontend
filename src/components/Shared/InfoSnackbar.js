import React from 'react';
import PropTypes from 'prop-types';

import { Snackbar, IconButton } from '@material-ui/core';
import CloseIcon from '@material-ui/icons/Close';

export function InfoSnackbar(props) {
  const { open, onClose, message } = props;
  return (
    <Snackbar
      anchorOrigin={{
        vertical: 'bottom',
        horizontal: 'left',
      }}
      open={open}
      autoHideDuration={5000}
      onClose={onClose}
      message={message}
      action={[
        <IconButton key="close" aria-label="Close" color="inherit" onClick={onClose}>
          <CloseIcon />
        </IconButton>,
      ]}
    />
  );
}

InfoSnackbar.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  message: PropTypes.string,
};

InfoSnackbar.defaultProps = {
  open: false,
  onClose: () => {},
  message: '',
};

export default InfoSnackbar;
