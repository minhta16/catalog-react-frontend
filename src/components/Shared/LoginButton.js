import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

import { Button, Avatar, Menu, MenuItem, Grid, Typography } from '@material-ui/core';
import PersonIcon from '@material-ui/icons/Person';
import ExitIcon from '@material-ui/icons/ExitToApp';

export class LoginButton extends Component {
  state = {
    anchorEl: undefined,
  };

  handleClickAvatar = (e) => {
    this.setState({
      anchorEl: e.target,
    });
  };

  handleCloseMenu = () => {
    this.setState({
      anchorEl: undefined,
    });
  };

  handleSignOut = () => {
    const { signOut } = this.props;
    signOut();
    this.handleCloseMenu();
  };

  render() {
    const { onClick, variant, color, currentUser } = this.props;
    const { anchorEl } = this.state;

    const menu = (
      <Menu
        id="simple-menu"
        anchorEl={anchorEl}
        keepMounted
        open={!!anchorEl}
        onClose={this.handleCloseMenu}
        getContentAnchorEl={null}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        transformOrigin={{ vertical: 'top', horizontal: 'center' }}
      >
        <MenuItem onClick={this.handleCloseMenu} component={Link} exact="true" to="/profile">
          <Grid container alignItems="center">
            <PersonIcon color="primary" style={{ margin: '0 0.5rem 0 0' }} />
            <Typography variant="body1">Profile</Typography>
          </Grid>
        </MenuItem>
        <MenuItem
          id="login-button-logout"
          onClick={this.handleSignOut}
          component={Link}
          exact="true"
          to="/"
        >
          <ExitIcon color="primary" style={{ margin: '0 0.5rem 0 0' }} />
          <Typography variant="body1">Logout</Typography>
        </MenuItem>
      </Menu>
    );
    return (
      <div>
        {Object.keys(currentUser).length ? (
          <Grid container alignItems="center">
            <Typography variant="body1">{`Welcome, ${currentUser.username}`}</Typography>
            <Avatar
              style={{ backgroundColor: '#fdb600', color: '#fff', margin: '0.5rem' }}
              onClick={this.handleClickAvatar}
            >
              {currentUser.username.charAt(0)}
            </Avatar>
            {menu}
          </Grid>
        ) : (
          <Button id="login-button" onClick={onClick} variant={variant} color={color}>
            Login
          </Button>
        )}
      </div>
    );
  }
}

LoginButton.propTypes = {
  onClick: PropTypes.func.isRequired,
  variant: PropTypes.string,
  color: PropTypes.string,
  currentUser: PropTypes.object.isRequired,
  signOut: PropTypes.func.isRequired,
};

LoginButton.defaultProps = {
  variant: 'contained',
  color: 'primary',
};

export default LoginButton;
