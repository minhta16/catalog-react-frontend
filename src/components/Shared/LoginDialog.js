import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  DialogTitle,
  Dialog,
  DialogContent,
  TextField,
  DialogActions,
  Button,
  Link,
} from '@material-ui/core';
import { Link as RouterLink } from 'react-router-dom';

export class LoginDialog extends Component {
  state = {
    username: '',
    password: '',
  };

  onClickSubmit = () => {
    const { onClick, onClose } = this.props;
    const { username, password } = this.state;
    onClick(username, password);
    onClose();
  };

  handleChange = (e) => {
    this.setState({
      [e.target.id]: e.target.value,
    });
  };

  render() {
    const { open, onClose } = this.props;
    const { username, password } = this.state;
    return (
      <Dialog open={open} onClose={onClose}>
        <DialogTitle>Login</DialogTitle>
        <DialogContent>
          <form autoComplete="off">
            <TextField
              required
              id="username"
              label="Username"
              value={username}
              onChange={this.handleChange}
              margin="normal"
              fullWidth
              variant="outlined"
            />
            <TextField
              required
              id="password"
              label="Password"
              value={password}
              onChange={this.handleChange}
              type="password"
              margin="normal"
              fullWidth
              variant="outlined"
            />
          </form>
          <Link onClick={onClose} component={RouterLink} exact="true" to="/register">
            Havenâ€™t got an account yet? Register here!
          </Link>
        </DialogContent>
        <DialogActions>
          <Button id="login-dialog-login" onClick={this.onClickSubmit} color="primary">
            Login
          </Button>
          <Button id="login-dialog-close" onClick={onClose}>
            Cancel
          </Button>
        </DialogActions>
      </Dialog>
    );
  }
}

LoginDialog.propTypes = {
  open: PropTypes.bool,
  onClose: PropTypes.func,
  onClick: PropTypes.func,
};

LoginDialog.defaultProps = {
  onClose: () => {},
  onClick: async (username, password) => username + password,
  open: true,
};
export default LoginDialog;
