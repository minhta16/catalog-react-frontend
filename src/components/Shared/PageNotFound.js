import React from 'react';
import { Link } from 'react-router-dom';
import { Typography, Button, Grid, Paper, Container } from '@material-ui/core';

const PageNotFound = () => (
  <Container maxWidth="lg">
    <Paper style={{ padding: '1rem' }}>
      <Grid container spacing={2} justify="center">
        <Grid item xs={12}>
          <Typography variant="h3">Page Not Found</Typography>
        </Grid>
        <Grid item xs={12}>
          <Typography variant="body1">
            The link you followed may be broken, or the page may have been removed.
          </Typography>
        </Grid>
        <Grid item xs={12}>
          <Button variant="contained" color="primary" component={Link} exact="true" to="/">
            Go back to Home
          </Button>
        </Grid>
      </Grid>
    </Paper>
  </Container>
);

export default PageNotFound;
