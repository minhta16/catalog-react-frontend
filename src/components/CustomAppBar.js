import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  AppBar, Toolbar, IconButton, Typography,
} from '@material-ui/core';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import { signIn as signInRedux } from 'references/redux/actions/users';
import LoginButton from './LoginButton';
import LoginDialog from './LoginDialog';

export class CustomAppBar extends Component {
  state = {
    open: false,
  };

  handleLoginClick = () => {
    const { open } = this.state;
    this.setState({
      open: !open,
    });
  };

  render() {
    const { open } = this.state;
    const { color, currentUser, signIn } = this.props;
    return (
      <AppBar color={color}>
        <Toolbar>
          <IconButton edge="start" color="inherit" component={Link} exact="true" to="/">
            <img src="/img/logo.svg" alt="site logo" />
          </IconButton>
          <Typography style={{ flexGrow: '1' }} />
          <LoginButton currentUser={currentUser} onClick={this.handleLoginClick} />
          <LoginDialog open={open} onClose={this.handleLoginClick} onClick={signIn} />
        </Toolbar>
      </AppBar>
    );
  }
}

CustomAppBar.propTypes = {
  color: PropTypes.string.isRequired,
  currentUser: PropTypes.object,
  signIn: PropTypes.func.isRequired,
};

CustomAppBar.defaultProps = {
  currentUser: { username: '', token: '' },
};

const mapStateToProps = state => ({
  currentUser: state.currentUser,
});

const mapDispatchToProps = dispatch => ({
  signIn: (username, password) => {
    dispatch(signInRedux(username, password));
  },
});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(CustomAppBar);
