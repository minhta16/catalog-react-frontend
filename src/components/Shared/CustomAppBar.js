import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { AppBar, Toolbar, IconButton, Typography } from '@material-ui/core';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import {
  signIn as signInRedux,
  signOut as signOutRedux,
  clearError as clearErrorRedux,
} from 'actions/users';
import { selectCurrentUser, selectLoginErrorMessage } from 'selectors/users';
import LoginButton from './LoginButton';
import LoginDialog from './LoginDialog';

export class CustomAppBar extends Component {
  state = {
    open: false,
  };

  handleOpenDialog = () => {
    this.setState({
      open: true,
    });
  };

  handleCloseDialog = () => {
    const { clearError } = this.props;
    clearError();
    this.setState({
      open: false,
    });
  };

  /**
   * close the login dialog if currentUser token exists, meaning that the user just logged in
   */
  componentDidUpdate = (prevProps) => {
    const { currentUser } = this.props;
    if (!prevProps.currentUser.token && currentUser.token) {
      this.handleCloseDialog();
    }
  };

  render() {
    const { open } = this.state;
    const { color, currentUser, signIn, signOut, errorMessage } = this.props;
    return (
      <AppBar color={color}>
        <Toolbar>
          <IconButton edge="start" color="inherit" component={Link} exact="true" to="/">
            <img src="/img/logo.svg" alt="site logo" />
          </IconButton>
          <Typography style={{ flexGrow: '1' }} />
          <LoginButton
            id="custom-app-bar-login-button"
            variant="contained"
            color="primary"
            currentUser={currentUser}
            onClick={this.handleOpenDialog}
            signOut={signOut}
          />
          <LoginDialog
            open={open}
            onClose={this.handleCloseDialog}
            onClick={signIn}
            errorMessage={errorMessage}
          />
        </Toolbar>
      </AppBar>
    );
  }
}

CustomAppBar.propTypes = {
  color: PropTypes.string.isRequired,
  currentUser: PropTypes.object,
  signIn: PropTypes.func.isRequired,
  signOut: PropTypes.func.isRequired,
  errorMessage: PropTypes.array,
  clearError: PropTypes.func.isRequired,
};

CustomAppBar.defaultProps = {
  currentUser: {},
  errorMessage: [],
};

export const mapSelectorToProps = (state) => ({
  currentUser: selectCurrentUser(state),
  errorMessage: selectLoginErrorMessage(state),
});

export const mapDispatchToProps = {
  signIn: signInRedux,
  signOut: signOutRedux,
  clearError: clearErrorRedux,
};

export default connect(
  mapSelectorToProps,
  mapDispatchToProps,
)(CustomAppBar);
