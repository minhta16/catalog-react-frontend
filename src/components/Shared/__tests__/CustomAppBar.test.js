/* eslint-disable no-undef */
import React from 'react';
import { shallow } from 'enzyme';
import { CustomAppBar } from 'components/Shared/CustomAppBar';
import LoginButton from 'components/Shared/LoginButton';
import LoginDialog from 'components/Shared/LoginDialog';

describe('components/Shared/CustomAppBar', () => {
  let wrapper;
  let props;
  // let warning;

  const update = () => {
    wrapper.update();
    wrapper.instance().forceUpdate();
  };

  const setup = () => {
    wrapper = shallow(<CustomAppBar {...props} />);
  };

  beforeEach(() => {
    props = {
      color: 'primary',
      signIn: jest.fn(),
      signOut: jest.fn(),
      currentUser: {
        username: 'minh',
        password: 'minh',
        token: 'abc',
      },
      errorMessage: '',
    };
  });

  it('should render correctly', () => {
    setup();
    expect(wrapper).toMatchSnapshot();
  });

  it('should handle handleOpenDialog() correctly', () => {
    setup();
    wrapper.instance().handleOpenDialog();
    expect(wrapper.state().open).toBe(true);
  });

  it('should handle handleCloseDialog() correctly', () => {
    setup();
    wrapper.instance().handleCloseDialog();
    expect(wrapper.state().open).toBe(false);
  });

  it('should call signIn when Dialog is clicked', () => {
    setup();
    const { signIn } = props;
    const loginDialog = wrapper.find(LoginDialog);
    loginDialog.simulate('click');
    expect(signIn).toHaveBeenCalled();
  });

  it('should handle componentDidUpdate correctly', () => {
    props.currentUser.token = 'meomeo';
    setup();
    wrapper.instance().handleCloseDialog = jest.fn();
    update();

    wrapper.instance().componentDidUpdate({
      currentUser: {
        token: 'still meomeo',
      },
    });
    expect(wrapper.instance().handleCloseDialog).not.toHaveBeenCalled();

    wrapper.instance().componentDidUpdate({
      currentUser: {
        token: '',
      },
    });
    expect(wrapper.instance().handleCloseDialog).toHaveBeenCalled();
  });

  it('should update state.open when handleLoginClick is called', () => {
    setup();
    const { open } = wrapper.state();
    const loginButton = wrapper.find(LoginButton);
    loginButton.simulate('click');
    expect(wrapper.state().open).toBe(!open);
  });
});
