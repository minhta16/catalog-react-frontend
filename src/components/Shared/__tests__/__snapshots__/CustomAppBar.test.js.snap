// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Shared/CustomAppBar should render correctly 1`] = `
<WithStyles(ForwardRef(AppBar))
  color="primary"
>
  <WithStyles(ForwardRef(Toolbar))>
    <WithStyles(ForwardRef(IconButton))
      color="inherit"
      component={[Function]}
      edge="start"
      exact="true"
      to="/"
    >
      <img
        alt="site logo"
        src="/img/logo.svg"
      />
    </WithStyles(ForwardRef(IconButton))>
    <WithStyles(ForwardRef(Typography))
      style={
        Object {
          "flexGrow": "1",
        }
      }
    />
    <LoginButton
      color="primary"
      currentUser={
        Object {
          "password": "minh",
          "token": "abc",
          "username": "minh",
        }
      }
      id="custom-app-bar-login-button"
      onClick={[Function]}
      signOut={[MockFunction]}
      variant="contained"
    />
    <LoginDialog
      errorMessage={Array []}
      onClick={[MockFunction]}
      onClose={[Function]}
      open={false}
    />
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
